File: projects/26-beehive-apiary-monitoring/backend/processor/transform.js
"use strict";

// site_id in the sort key keeps two apiaries' same-sensor readings in one flush cycle from colliding on the primary key.
function composeHiveSortKey(windowEnd, apiaryId) {
  return `${windowEnd}#${apiaryId}`;
}

function toHiveReadingItem(rawWindow) {
  const window = typeof rawWindow === "string" ? JSON.parse(rawWindow) : rawWindow;
  const apiaryId = window.site_id || "apiary-a";
  return {
    sensor_type: window.sensor_type,
    sort_key: composeHiveSortKey(window.window_end, apiaryId),
    window_end: window.window_end,
    window_start: window.window_start,
    site_id: apiaryId,
    unit: window.unit || "",
    count: window.count,
    min: window.min,
    max: window.max,
    avg: window.avg,
    latest: window.latest,
    alerts: window.alerts || [],
  };
}

module.exports = { toHiveReadingItem, composeHiveSortKey };

File: projects/14-smart-parking-management/backend/processor/transform.py
import json


def to_item(message_body):
    """Transform a raw SQS message body into the flat DynamoDB record; sort_key is window_end#site_id to avoid per-lot collisions."""
    data = json.loads(message_body) if isinstance(message_body, str) else message_body
    site_id = data.get("site_id", "lot-a")
    return {
        "sensor_type": data["sensor_type"],
        "sort_key": f"{data['window_end']}#{site_id}",
        "window_start": data["window_start"],
        "window_end": data["window_end"],
        "site_id": site_id,
        "unit": data.get("unit", ""),
        "count": data["count"],
        "min": data["min"],
        "max": data["max"],
        "avg": data["avg"],
        "latest": data["latest"],
        "alerts": data.get("alerts", []),
    }
